### PR TITLE
Add support for storage bucket recovery

### LINKED ISSUE #11704: lxd recover: add support for storage buckets
...this is general, so distribution & other environment information have been omitted.

# Issue description

lxd recover currently does not include storage buckets, that would be nice to have.

and it could scan the filesystem and automatically restore all buckets or a selection.

on a side note, this would be nice to have for storage pools and volumes too, so the user need not remember or need to mount and explore the filesystem manually to recall the names.

### PR DESCRIPTION
**Pull request by MusicDin**
Fixes #11704

### CODE REVIEW

**Comment by tomponline** [comment]
@MusicDin this can be rebased now.

**Comment by tomponline** [comment]
This can be rebased now.

**Review by tomponline — CHANGES_REQUESTED** [review]
Looks good.

Are you able to update the recovery tests with bucket support?

https://github.com/canonical/lxd/blob/2a718d05d8e0abb24e527fad2b5836dbbbf257ae/test/suites/backup.sh#L1

**Comment by tomponline** [comment]
> Any thoughts on that?

Yes I think that makes sense, as not all storage drivers support buckets and we dont want to break recovery in that case.

**Review by tomponline — COMMENTED** [review]
Please let me know when you would like another review  Thanks

**Comment by MusicDin** [comment]
The test for docs is failing:
```
(installing: line   48) broken    https://snapcraft.io/docs/core/install - 500 Server Error:  for url: https://snapcraft.io/docs/installing-snapd
```

While I have not changed anything in the docs and both URLs are working, I will assume it was a temporary error.

---

@tomponline @monstermunchkin @roosterfish Can you please review the code? And if possible, please check the test (https://github.com/canonical/lxd/pull/11883/commits/30bc0058c62eb974589bd84111041fc585775150) as this is my first test, so I could benefit from comments or any "tips and tricks"?

**Comment by tomponline** [comment]
> The test for docs is failing:
> 
> ```
> (installing: line   48) broken    https://snapcraft.io/docs/core/install - 500 Server Error:  for url: https://snapcraft.io/docs/installing-snapd
> ```
> 
> While I have not changed anything in the docs and both URLs are working, I will assume it was a temporary error (also the `make doc-linkcheck` passes locally).
> 
> @tomponline @monstermunchkin @roosterfish Can you please review the code? And if possible, please check the test ([30bc005](https://github.com/canonical/lxd/commit/30bc0058c62eb974589bd84111041fc585775150)) as this is my first test, so I could benefit from comments or any "tips and tricks"?

Thats just the snapstore failing

File: lxd/api_internal_recover.go [hunk 1]
```diff
@@ -221,6 +223,10 @@ func internalRecoverScan(s *state.State, userPools []api.StoragePoolsPost, valid
 		// Get list of unknown volumes on pool.
 		poolProjectVols, err := pool.ListUnknownVolumes(nil)
 		if err != nil {
+			if errors.Is(err, drivers.ErrNotSupported) {
```
Comment (tomponline) [inline lxd/api_internal_recover.go:226]: This should be a separate commit as isnt anything to do with bucket recovery.

File: lxd/api_internal_recover.go [hunk 2]
```diff
@@ -19,6 +20,7 @@ import (
 	"github.com/canonical/lxd/lxd/revert"
 	"github.com/canonical/lxd/lxd/state"
 	storagePools "github.com/canonical/lxd/lxd/storage"
+	"github.com/canonical/lxd/lxd/storage/drivers"
```
Comment (tomponline) [inline lxd/api_internal_recover.go:23]: Please can you import this using the convention:

```
	storageDrivers "github.com/canonical/lxd/lxd/storage/drivers"
```

File: lxd/storage/backend_lxd.go [hunk 3]
```diff
@@ -3731,19 +3735,135 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
...
+		}
+
+		svcAccountInfo, err := adminClient.InfoServiceAccount(ctx, creds.AccessKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// If service account policy equals to the admin bucket policy, set bucket key role to admin.
+		if comparePolicy(string(adminPolicy), svcAccountInfo.Policy) {
+			key.Role = "admin"
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:3833]: What do you think about defining some constant(s) that we can use here and in where we create/manage the key roles so we can easily see in the future what parts of the code base deal with bucket key roles.

File: lxd/storage/backend_lxd.go [hunk 4]
```diff
@@ -3731,19 +3735,135 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
...
+		revert.Add(func() { _ = b.state.DB.Cluster.DeleteStoragePoolBucketKey(context.TODO(), bucketID, keyID) })
+	}
+
 	cleanup := revert.Clone().Fail
 	revert.Success()
 	return cleanup, err
 }
 
+// comparePolicy compares S3 policies and returns true if they are strictly equal.
+func comparePolicy(policyA, policyB string) bool {
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:3850]: Can we move this to storage/utils.go and call it something clearer, like `minioPolicyCompare`?

Comment (tomponline) [reply lxd/storage/backend_lxd.go:3850]: We dont use argument type compaction (not sure if thats the right terminology), so please can the function signature be:

```
comparePolicy(policyA string, policyB string) bool
```

**Review by tomponline — CHANGES_REQUESTED** [review]
This is coming along very nicely, I like the new minio policy parsing stuff. :)

A few minor nits to address and then we're ready to merge.

File: lxd/storage/backend_lxd.go (hunk 3 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:3833]: Agree, will do that.

File: lxd/storage/backend_lxd.go (hunk 4 above)
Comment (tomponline) [reply lxd/storage/backend_lxd.go:3850]: Yep sounds good! Does this only work for minio though or is it a general s3 thing?

File: lxd/storage/backend_lxd.go [hunk 5]
```diff
@@ -3696,6 +3696,54 @@ func (b *lxdBackend) DeleteBucket(projectName string, bucketName string, op *ope
...
+		bucketConfig[k] = v
+	}
+
+	bucket := &api.StorageBucketsPost{
+		Name:             poolVol.Bucket.Name,
+		StorageBucketPut: poolVol.Bucket.StorageBucketPut,
+	}
+
+	// Validate config and create database entry for restored bucket.
+	bucketID, err := BucketDBCreate(context.TODO(), b, projectName, true, bucket)
```
Comment (monstermunchkin) [inline lxd/storage/backend_lxd.go:3730]:
```suggestion
	bucketID, err := BucketDBCreate(b.state.ShutdownCtx, b, projectName, true, bucket)
```

File: lxd/storage/backend_lxd.go [hunk 6]
```diff
@@ -3696,6 +3696,54 @@ func (b *lxdBackend) DeleteBucket(projectName string, bucketName string, op *ope
...
+	if err != nil {
+		return nil, err
+	}
+
+	revert.Add(func() { _ = BucketDBDelete(context.TODO(), b, bucketID) })
```
Comment (monstermunchkin) [inline lxd/storage/backend_lxd.go:3735]:
```suggestion
	revert.Add(func() { _ = BucketDBDelete(b.state.ShutdownCtx, b, bucketID) })
```

File: lxd/storage/backend_lxd.go [hunk 7]
```diff
@@ -3696,6 +3696,54 @@ func (b *lxdBackend) DeleteBucket(projectName string, bucketName string, op *ope
...
+	storageBucket := b.GetVolume(drivers.VolumeTypeBucket, drivers.ContentTypeFS, storageBucketName, bucketConfig)
+
+	err = b.driver.ValidateVolume(storageBucket, false)
+	if err != nil {
+		return nil, err
+	}
+
+	cleanup := revert.Clone().Fail
+	revert.Success()
+	return cleanup, err
```
Comment (monstermunchkin) [inline lxd/storage/backend_lxd.go:3846]:
```suggestion
	return cleanup, nil
```

File: lxd/storage/backend_lxd.go (hunk 4 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:3850]: As far as I know, it is generic S3 policy.

Just as a note, comparing policies like this will work only for the roles that are predefined in the `lxd/storage/s3/policy.go`. Since any difference in the policy would result in invalid comparison. However, since the policy is managed exclusively by LXD I think this should not be a problem (at least from a perspective of bucket recovery).

File: lxd/storage/backend_lxd.go [hunk 8]
```diff
@@ -3731,19 +3735,135 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
...
+		return nil, err
+	}
+
+	// Initialize minio client object.
+	adminClient, err := minioProc.AdminClient()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, ctxCancel := context.WithTimeout(context.TODO(), time.Duration(time.Second*30))
```
Comment (monstermunchkin) [inline lxd/storage/backend_lxd.go:3760]:
```suggestion
	ctx, ctxCancel := context.WithTimeout(b.state.ShutdownCtx, time.Duration(time.Second*30))
```

File: lxd/storage/backend_lxd.go [hunk 9]
```diff
@@ -3731,19 +3735,135 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
...
+		if err != nil {
+			return nil, err
+		}
+
+		// If service account policy equals to the admin bucket policy, set bucket key role to admin.
+		if comparePolicy(string(adminPolicy), svcAccountInfo.Policy) {
+			key.Role = "admin"
+		}
+
+		keyID, err := b.state.DB.Cluster.CreateStoragePoolBucketKey(context.TODO(), bucketID, key)
```
Comment (monstermunchkin) [inline lxd/storage/backend_lxd.go:3836]:
```suggestion
		keyID, err := b.state.DB.Cluster.CreateStoragePoolBucketKey(b.state.ShutdownCtx, bucketID, key)
```

File: lxd/storage/backend_lxd.go [hunk 10]
```diff
@@ -3731,19 +3735,135 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
...
+		if err != nil {
+			return nil, err
+		}
+
+		revert.Add(func() { _ = b.state.DB.Cluster.DeleteStoragePoolBucketKey(context.TODO(), bucketID, keyID) })
```
Comment (monstermunchkin) [inline lxd/storage/backend_lxd.go:3841]:
```suggestion
		revert.Add(func() { _ = b.state.DB.Cluster.DeleteStoragePoolBucketKey(b.state.ShutdownCtx, bucketID, keyID) })
```

**Review by monstermunchkin — CHANGES_REQUESTED** [review]

**Comment by MusicDin** [comment]
Hi @tomponline, can you take another look at the latest changes in [`lxd/storage/s3/policy.go`](https://github.com/canonical/lxd/pull/11883/commits/89f9eef7faf109da7946f6381fb2c2f4f9520094)?

File: lxd/storage/backend_lxd.go [hunk 11]
```diff
@@ -3723,25 +3726,118 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
 	}
 
 	// Validate config and create database entry for restored bucket.
-	bucketID, err := BucketDBCreate(context.TODO(), b, projectName, true, bucket)
+	bucketID, err := BucketDBCreate(b.state.ShutdownCtx, b, projectName, true, bucket)
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:3729]: These changes to passing in context are unrelated to the commit message and should be in the earlier commit that added this function.

File: lxd/storage/backend_lxd.go [hunk 12]
```diff
@@ -3723,25 +3726,118 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
...
 
 	err = b.driver.ValidateVolume(storageBucket, false)
 	if err != nil {
 		return nil, err
 	}
 
+	// Recover bucket keys.
+
+	// Start minio process.
+	minioProc, err := b.ActivateBucket(projectName, bucket.Name, op)
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:3748]: This is a problem because `ImportBucket` should (at least someday) need to work with cephobject pools that dont use minio. You can see the other invocations of b.ActivateBucket are always conditional on the bucket being memberSpecific (i.e not a remote ceph bucket).

It sounds like we need a change to ensure that ImportBucket handles (or at least cleanly passes over) recovering from a cephobject pool, and doesn't assume all buckets are minio based.

File: lxd/storage/backend_lxd.go [hunk 13]
```diff
@@ -3723,25 +3726,118 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
...
+			return nil, err
+		}
+
+		revert.Add(func() { _ = b.state.DB.Cluster.DeleteStoragePoolBucketKey(b.state.ShutdownCtx, bucketID, keyID) })
+	}
+
 	cleanup := revert.Clone().Fail
 	revert.Success()
-	return cleanup, err
+	return cleanup, nil
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:3840]: this isnt related to the commit message

File: lxd/storage/backend_lxd.go [hunk 14]
```diff
@@ -3723,25 +3726,118 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
...
+	// Initialize minio client object.
+	adminClient, err := minioProc.AdminClient()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, ctxCancel := context.WithTimeout(b.state.ShutdownCtx, time.Duration(time.Second*30))
+	defer ctxCancel()
+
+	// Export IAM data (response is ZIP file).
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:3762]: Can this section below be moved into its own function on backendLXD called something like `recoverMinIOKeys`

File: lxd/storage/backend_lxd.go (hunk 12 above)
Comment (tomponline) [reply lxd/storage/backend_lxd.go]: Didnt you already add a check that skips if ErrNotSupported returned? Could you just return that in this case?

Comment (tomponline) [reply lxd/storage/backend_lxd.go]: Ah no that was in ListUnknownVolumes. I think returning an error as you propose is best for now, as it shouldn't every actually get there because it will never be in the unknown volume list.

File: lxd/storage/backend_lxd.go [hunk 15]
```diff
@@ -3696,6 +3699,171 @@ func (b *lxdBackend) DeleteBucket(projectName string, bucketName string, op *ope
...
+		for _, key := range keys {
+			keyID, err := b.state.DB.Cluster.CreateStoragePoolBucketKey(b.state.ShutdownCtx, bucketID, key)
+			if err != nil {
+				return nil, err
+			}
+
+			revert.Add(func() { _ = b.state.DB.Cluster.DeleteStoragePoolBucketKey(b.state.ShutdownCtx, bucketID, keyID) })
+		}
+	} else {
+		return nil, fmt.Errorf("Imporiting buckets from a remote storage is not supported")
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:3766]: Typo

**Review by tomponline — COMMENTED** [review]
Much cleaner thanks. One typo to go :)

File: lxd/storage/backend_lxd.go (hunk 12 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go]: Yes, of course.

To clarify the check we have added previously:
In the `lxd/api_internal_recover` a function `lxd_backend#ListUnknownVolumes()` is called which calls `driver#ListVolumes()`. Since Ceph object storage does not support volumes, the recovery failed (this is currently a *bug* on master branch).

File: lxd/api_internal_recover.go [hunk 16]
```diff
@@ -417,8 +421,8 @@ func internalRecoverScan(s *state.State, userPools []api.StoragePoolsPost, valid
 
 			// Recover unknown instance volumes.
 			for _, poolVol := range poolVols {
-				if poolVol.Container == nil && poolVol.Volume != nil {
-					continue // Skip custom volumes and invalid volumes.
+				if poolVol.Container == nil && poolVol.Volume != nil || poolVol.Bucket != nil {
```
Comment (monstermunchkin) [inline lxd/api_internal_recover.go:430]: Just double checking. Is this correct or should it be?
```go
if poolVol.Container == nil && (poolVol.Volume != nil || poolVol.Bucket != nil) {
```

Comment (MusicDin) [reply lxd/api_internal_recover.go:430]: If `poolVol.Container != nil && poolVol.Volume != nil` => instance volume
If `poolVol.Container == nil && poolVol.Volume != nil` => Custom volume
If `poolVol.Bucket != nil` => Bucket volume

So this condition is meant as:
```go
if (poolVol.Container == nil && poolVol.Volume != nil) || poolVol.Bucket != nil {
//  ^ Is custom volume                                    ^ Is bucket
```

In this specific case, even `poolVol.Container == nil && (poolVol.Volume != nil || poolVol.Bucket != nil)` would work, since `poolVol.Container == nil` is true for both custom volumes, and buckets.

Which means that for clarity, I would prefer your approach.

**Comment by MusicDin** [comment]
If there are no other comments, I think this is ready to be merged.

**Comment by tomponline** [comment]
Thanks!
